Fix #71890: Add support for crc32c Castagnoli's polynomial.

bramp opened the pull request:
This variant of crc32 is heavily used by storage systems, such as iSCSI, SCTP,
Btrfs, ext4, and is increasingly being used in APIs (such as Google Cloud
Storage, and Apache Kafka).

nikic commented:
Implementation looks fine to me. Are there any official test vectors that can be included to check that the output is the right one?

cmb69 commented:
Should this really target PHP-7.2? After all, I don't think it's a bugfix.

KalleZ commented:
I think this should target 7.4 if anything. From a quick glance it looks great, if we could have a few more conforming tests like @nikic mentioned, it would be great.

(Note: some bugs at the bugsweb is actually a feature request so we format it so)

bramp commented:
Thanks for the quick review.

I guess I misread the instructions, as it suggested rebasing against the earliest supported version (which looked to be 7.2). Happy to rebase against `master` instead. However, it would be really useful (to me) if this was in 7.2. Is there anyway I can achieve that?

I looked for testing vectors, but sadly there seems to be no golden set. The [original academic paper](https://ieeexplore.ieee.org/document/231911) lists none, the [iSCSI RFC](https://tools.ietf.org/html/rfc3720#appendix-B.4) has 
 has a few. The best I could find is [Go's implementation](https://golang.org/src/hash/crc32/crc32_test.go) which has ~30, but they are all testing ASCII strings.  

So I can include what I've found in other implementations. Or an alternative. I write a quick program in another language (say Go), and generate samples that include all bytes (0-255) and use that to generate a canonical list.

I have a day job, so I'll make the requested changes over the next day or so. Thanks.

nikic commented:
It's probably a bit late for 7.2, but I think we can land this on 7.3 under the "small self-contained feature" provision, if @cmb69 agrees.

I think it would be good to include both those Go test vectors as well as the coverage for all bytes. If we have that shouldn't be much opportunity to go wrong :)

cmb69 commented:
> I guess I misread the instructions, as it suggested rebasing against the earliest supported version (which looked to be 7.2). Happy to rebase against master instead. However, it would be really useful (to me) if this was in 7.2. Is there anyway I can achieve that?

7.2 would be appropriate if this was a bug; I don't think the lack of any particular hash algo is a bug, though. If you want to see this enhancement earlier than in PHP-7.4, I suggest you write to internals@lists.php.net; maybe there is some consensus about adding it to PHP-7.2 (or maybe 7.3).

bramp commented:
I've rebased to master, and I've added additional tests (using Go to generate the golden hash values). For now I'll stick with 7.4, and I'll use a crc32c implementation written in php for earlier versions.

Please take a look.

KalleZ commented:
Looks great, thank you for your work! I suggest you hit a mail to internals@ to decide on what version to target, as this is a small and self contained feature, it should be fast to decide

bramp commented:
BTW Sent email to the internals@ list, awaiting a decision on what version(s) to merge into.

nikic commented:
It seems like the sentiment on list is to not merge into release branches, so I went ahead and merged this as https://github.com/php/php-src/commit/c79ce48ddb2125a3d19723e0fbca1bd770bacab7 into 7.4. We can always backport if necessary (the patch would have to be slightly different for older versions anyway).

Thanks for adding this functionality :)

bramp commented:
Thanks!
